File: concept-task/input-data/format-CRP-input.R
library(ggplot2)
library(ggbiplot)
library(dplyr)
library(tidyr)
library(lme4)
library(lmerTest)
library(Rtsne)
library(RColorBrewer)
library(colorRamps)
library(patchwork)
library(ggpubr)
library(scales)
library(lsa)
library(stringr)

getwd()

data <- read.csv("./output-data/CHANGETHIS.csv", header = FALSE) %>%
  filter(!row_number() %in% c(1, 2, 4)) # remove first row

# rename columns
colnames(data) <- c("ID", "Prompting_condition", "Experiment", "Concept", "Choice1", "Choice2", "Response1", "Response2")

### Data formatting
# capital case response columns
data$Response1 <- str_to_title(data$Response1)
data$Response2 <- str_to_title(data$Response2)

# recode responses and factorize
data$Concept <- as.factor(data$Concept)
data$Concept <- recode(data$Concept, " a chicken" = "chicken", " a dolphin" = "dolphin", " a finch" = "finch",
                                     " a penguin" = "penguin", " a robin" = "robin", " a salmon" = "salmon",
                                     " a seal" = "seal", " a whale" = "whale", " an ostrich" = "ostrich",
                                     " an eagle" = "eagle", " Abraham Lincoln" = "Lincoln", " Barack Obama" = "Obama",
                                     " Bernie Sanders" = "Sanders", " Donald Trump" = "Trump", " Elizabeth Warren" = "Warren",
                                     " George W. Bush" = "Bush", " Hillary Clinton" = "Clinton", " Joe Biden" = "Biden",
                                     " Richard Nixon" = "Nixon", " Ronald Reagan" = "Reagan")
# changes the order of the factors for visualization
data$Concept <- factor(data$Concept, levels(data$Concept)[c(3, 5, 1, 10, 11, 4, 6, 7, 2, 8, 9, 12:20)])

# factorize Experiment column (animals vs. politicians)
data$Experiment <- as.factor(data$Experiment)

# make new column for question by combining the two choice options and factorize
data$Question <- paste(as.character(data$Choice1), as.character(data$Choice2))
data$Question <- as.factor(data$Question)

# "We coded each participant’s responses to a single word as a binary vector, corresponding to theforced-choice similarity rating between every other pair of items"
# encode this based on the model's first response
data$ChoiceNumber <- ifelse(as.character(data$Response1) == as.character(data$Choice1), 0, 1)

# If verifying human data, skip above and load in and format data
data <- read.csv("../../input-data/concept-task/Responses.csv", header = FALSE) %>%
  filter(!row_number() %in% c(1, 2, 4)) %>%
  rename_with(~c("i", "Concept", "ID", "Question", "ChoiceNumber"))

### Calculate participant reliability
# .rs.restartR()
tmp <- data %>% 
  group_by(ID, Question, Concept) %>% 
  summarise(Reliability = mean(ChoiceNumber))

tmp$Reliability <- ifelse(tmp$Reliability == 0, 1, ifelse(tmp$Reliability == 1, 1, 0))
data <- merge(data, tmp)

reliability <- data.frame(data$Reliability)
reliability$ID <- data$ID

reliability$lower <- ifelse(data$Reliability,
                            prop.test(sum(data$Reliability), length(data$Reliability))$conf.int[1],
                            prop.test(sum(data$Reliability == 0), length(data$Reliability))$conf.int[1])

reliability$upper <- ifelse(data$Reliability,
                            prop.test(sum(data$Reliability), length(data$Reliability))$conf.int[2],
                            prop.test(sum(data$Reliability == 0), length(data$Reliability))$conf.int[2])

tmp <- reliability %>% group_by(ID) %>% summarise(reliabilityPercent = mean(data.Reliability))
data <- merge(data, tmp)
reliability <- merge(reliability, tmp)

reliability$data.Reliability <- factor(reliability$data.Reliability, levels = c(0, 1), labels = c("Not Reliable", "Reliable"))

### Calculate Intersubject Reliability
tmp <- data %>% group_by(Question, Concept) %>% summarise(QuestionReliability = mean(ChoiceNumber))
tmp <- tmp %>% group_by(Concept) %>% mutate(ConceptReliability = mean(QuestionReliability))
mean(tmp$QuestionReliability)


### Save formatted data
alpha <- .16
probs <- rbeta(1000000, alpha, alpha)
agree <- 1 != rbinom(1000000, 2, probs)
prop.table(table(agree))

mean(data$Reliability) # alpha = .16
mean(subset(data, Experiment %in% c("Animals"))$Reliability) # alpha = .14
mean(subset(data, Experiment %in% c("Politicians"))$Reliability) #alpha = .18

# Write people's responses to CSV for use in the CRP model
responses <- data %>% distinct(ID, Question, Concept, .keep_all = TRUE)
responses <- responses %>% select(Concept, ID, Question, ChoiceNumber)
write.csv(responses, "./output-data/CHANGETHIS.csv")

File: concept-task/input-data/format-CRP-input.py
import pandas as pd
import numpy as np
from scipy import stats
from sklearn.manifold import TSNE
import matplotlib.pyplot as plt
import seaborn as sns
from functools import reduce
import re
import warnings
warnings.simplefilter(action='ignore', category=FutureWarning)

"""REPLICATION OF DATA FORMATTING SECTION OF ANALYSIS.R"""


def format_data_for_CRP(data):
    # rename columns via mapping dictionary
    column_mapping = {
        "subject_num": "ID", "model_name": "Model", "prompt": "Prompting_condition", "temperature": "Temperature",
        "concept_category": "Experiment", "target": "Concept", "choice1": "Choice1", "choice2": "Choice2",
        "answer1": "Response1", "answer2": "Response2"
    }

    data.rename(columns=column_mapping, inplace=True)

    # Recode responses and factorize
    concept_mapping = {
        " a chicken": "chicken", " a dolphin": "dolphin", " a finch": "finch",
        " a penguin": "penguin", " a robin": "robin", " a salmon": "salmon",
        " a seal": "seal", " a whale": "whale", " an ostrich": "ostrich",
        " an eagle": "eagle", " Abraham Lincoln": "Lincoln", " Barack Obama": "Obama",
        " Bernie Sanders": "Sanders", " Donald Trump": "Trump", " Elizabeth Warren": "Warren",
        " George W. Bush": "Bush", " Hillary Clinton": "Clinton", " Joe Biden": "Biden",
        " Richard Nixon": "Nixon", " Ronald Reagan": "Reagan"
    }

    data['Concept'] = data['Concept'].replace(concept_mapping)

    # Create Question column
    data['Question'] = data['Choice1'] + " " + data['Choice2']
    data['Question'] = data['Question'].astype('category')

    # Encode choice number based on Response1 column
    data['ChoiceNumber'] = (data['Response1'] != data['Choice1']).astype(int)

    return data

"""Takes in formatted data for CRP model and calculates participant reliability metrics"""
def calculate_participant_reliability(data):
    # gets the average of the ChoiceNumber for each subject, question, and concept --> if the average is 0 or 1, the response is reliable because both responses are the same
    tmp = data.groupby(['ID', 'Question', 'Concept']).agg({'ChoiceNumber': 'mean'}).reset_index()
    # check that the average is 0 or 1, if it is then Reliability is True, otherwise False
    tmp['Reliability'] = (tmp['ChoiceNumber'] == 0) | (tmp['ChoiceNumber'] == 1)

    data = pd.merge(data, tmp[['ID', 'Question', 'Concept', 'Reliability']], on=['ID', 'Question', 'Concept'])

    # Create reliability dataframe
    reliability = pd.DataFrame({'data_Reliability': data['Reliability'], 'ID': data['ID']})

    # Calculate confidence intervals
    def wilson_score_interval(n, p, z=1.96):
        denominator = 1 + z**2/n
        center_adjusted_probability = p + z*z / (2*n)
        adjusted_standard_deviation = z * np.sqrt((p*(1 - p) + z*z/(4*n)) / n)
        
        lower_bound = (center_adjusted_probability - adjusted_standard_deviation) / denominator
        upper_bound = (center_adjusted_probability + adjusted_standard_deviation) / denominator
        
        return (lower_bound, upper_bound)

    n = len(data)
    p = data['Reliability'].mean()
    lower, upper = wilson_score_interval(n, p)
    reliability['lower'] = lower
    reliability['upper'] = upper

    # Calculate reliability percent
    tmp = reliability.groupby('ID')['data_Reliability'].mean().reset_index()
    tmp.columns = ['ID', 'reliabilityPercent']
    data = pd.merge(data, tmp, on='ID')
    
    reliability = pd.merge(reliability, tmp, on='ID')

    reliability['data_Reliability'] = reliability['data_Reliability'].map({False: "Not Reliable", True: "Reliable"})

    return data, reliability


"""Takes in formatted data for CRP model and calculates intersubject reliability metrics"""
def calculate_intersubject_reliability(data):
    tmp = data.groupby(['Question', 'Concept']).agg({'ChoiceNumber': 'mean'}).reset_index()
    tmp.columns = ['Question', 'Concept', 'QuestionReliability']

    tmp['ConceptReliability'] = tmp.groupby('Concept')['QuestionReliability'].transform('mean')
    print(f"Mean Question Reliability: {tmp['QuestionReliability'].mean().round(4)}")

    return tmp

def verify_human_data(path_to_human_data):
    # load human data
    formattedData = pd.read_csv(path_to_human_data)

    formattedData, participant_reliability = calculate_participant_reliability(formattedData)
    intersubject_reliability = calculate_intersubject_reliability(formattedData)

    # print(participant_reliability)
    # print(intersubject_reliability)

    # Calculate agreement
    alpha = 0.16
    probs = np.random.beta(alpha, alpha, 1000000)
    agree = 1 != np.random.binomial(2, probs)
    print(f"Agreement proportion: {np.mean(agree)}")

    # count number of each value in the Reliability column
    print(formattedData['Reliability'].value_counts())
    print(f"Mean Reliability: {formattedData['Reliability'].mean()}")


model_names = ["openchat", "starling", "mistral-instruct", "zephyr-mistral", "gemma-instruct", "zephyr-gemma", "llama2", "llama2-chat", "tulu", "tulu-dpo"]
param_combos = [["none", "default"], ["none", "1.5"], ["none", "2.0"], ["random", "default"], ["nonsense", "default"], ["identity", "default"]]
categories = ["animals", "politicians"]
data_dir = "../../output-data/concept-task/"

# first, verify code on human data
verify_human_data("../../input-data/concept-task/Responses.csv")

for model in model_names:
    for param_combo in param_combos:
        # read in pickle file
        data = pd.read_pickle(f"{data_dir}concept-{model}.pickle")
        prompt = param_combo[0]
        temperature = param_combo[1]

        print(f"Model: {model}, Prompt: {prompt}, Temperature: {temperature}")

        # select data for the current model, category, and parameter combination
        data = data[(data['prompt'] == prompt) & (data['temperature'] == temperature)]
        # select columns
        data = data[['subject_num', 'model_name', 'prompt', 'temperature', 'concept_category', 'target', 'choice1', 'choice2', 'answer1', 'answer2']]
        # remove rows where either answer is -1
        data = data[(data['answer1'] != -1) & (data['answer2'] != -1)]

        # # verify that there's no data where target is equal to either choice1 or choice2
        # print(data[(data['target'] == data['choice1']) | (data['target'] == data['choice2'])])

        formattedData = format_data_for_CRP(data)
        responses = formattedData.drop_duplicates(subset=['ID', 'Question', 'Concept'])
        responses = responses[['Concept', 'ID', 'Question', 'ChoiceNumber']]
        responses.to_csv(f"{model}-prompt={prompt}-temp={temperature}-forCRP.csv", index=False)

        formattedData, participant_reliability = calculate_participant_reliability(formattedData)
        intersubject_reliability = calculate_intersubject_reliability(formattedData)

        print(intersubject_reliability)

        # Calculate agreement
        alpha = 0.16
        probs = np.random.beta(alpha, alpha, 1000000)
        agree = 1 != np.random.binomial(2, probs)
        print(f"Agreement proportion: {np.mean(agree)}")

        print(f"Mean Reliability: {formattedData['Reliability'].mean()}")
        print(f"Mean Reliability (Animals): {formattedData[formattedData['Experiment'] == 'animals']['Reliability'].mean()}")
        print(f"Mean Reliability (Politicians): {formattedData[formattedData['Experiment'] == 'politicians']['Reliability'].mean()}")

        print("-" * 50)

